fuzz: compact block harness

Crypt-iQ opened the pull request:
Adds a fuzz harness for testing compact blocks, similar to `process_message(s)`. It can make (in)valid blocks, reconstruct blocks with in-mempool txns, mark peers as HB, and has high stability in AFL++ (~98-99%).

Coverage is [here](https://crypt-iq.github.io/fuzz_coverage_reports/cmpctblock-aflpp-inputs-09182025/) (look in `src/blockencodings.cpp`, relevant compact block bits in `src/net_processing.cpp`).

dergoegge commented:
> Is my understanding correct that AFL will run init (create the static dir), then fork into several fuzz processes, which run for N iterations and then shut down and delete the static dir? (On the next re-fork, it will fail to find the static dir)?

We removed the `__AFL_INIT` call, so it will actually fork prior to `init`, so I think all that's needed is to randomize the static dir name as you suggested as well.

Crypt-iQ commented:
I was able to get AFL++ multi-core fuzzing to work by adding a static `FuzzedDirectoryWrapper` instance that deletes the passed path in its destructor as well as using a random element to each statically named datadir. Each instance is in the high 98-99% stability and does not crash after hitting 100,000 iterations (yay). I was not able to use multiple `TestingSetup`s at the same time without causing several panics due to assertions about globals. I will fix the lint, tidy jobs and see if the determinist-fuzz-coverage issues still pop up.

Crypt-iQ commented:
There is non-determinism [here](https://github.com/bitcoin/bitcoin/blob/c0894a0a2be032cd9a5d5945643689230ab10255/src/node/blockstorage.cpp#L490-L493) because `m_dirty_blockindex` is `std::set<CBlockIndex*>` and sorts based on the pointer. This can be fixed by adding a function object that compares the block hashes. However, I did not know if this should be added just for fuzz code and I have not run benchmarks.

After patching the above non-determinism locally, there is also non-determinism in the scheduler because this fuzz test uses the scheduler. I am guessing this is due to the `RegisterValidationInterface` call in the fuzz test notifying when transactions are being added or removed, and blocks being mined.

dergoegge reviewed (changes requested):
Did a first pass, overall approach looks good to me

Crypt-iQ commented on src/test/fuzz/cmpctblock.cpp:
```diff
@@ -0,0 +1,421 @@
+namespace {
+
+//! Fee each created tx will pay.
+const CAmount amount_fee = 1000;
```
Thanks, done.

Crypt-iQ commented on src/test/fuzz/cmpctblock.cpp:
```diff
@@ -0,0 +1,421 @@
+
+                // Choose an existing block and send a HEADERS message for it.
+                size_t index = fuzzed_data_provider.ConsumeIntegralInRange<size_t>(0, num_blocks - 1);
+                CBlock block = *info[index].block;
```
Ah nice catch, I completely missed this. Done.

Crypt-iQ commented on src/test/fuzz/cmpctblock.cpp:
```diff
@@ -0,0 +1,421 @@
+                    ++num_erased;
+                }
+
+                CBlockHeaderAndShortTxIDs baseCmpctBlock = cmpctBlock;
```
Thanks, I believe I've fixed all of the camel case.

marcofleon commented on src/test/fuzz/cmpctblock.cpp:
```diff
@@ -0,0 +1,422 @@
+
+        CBlockIndex* pindexPrev{WITH_LOCK(::cs_main, return setup->m_node.chainman->m_blockman.LookupBlockIndex(prev))};
+        setup->m_node.chainman->GenerateCoinbaseCommitment(*block, pindexPrev);
+        FinalizeHeader(header, *setup->m_node.chainman);
```
Shouldn't `*block` be passed in here instead of `header`? Right now, I think the block just ends up with the random nonce from when the header was created. Although this still works half the time with the simplifed PoW check. Also, calling `GetHash()` on the header includes the merkle root, so finalizing the header should be after we calculate the merkle root below.

stringintech commented on src/test/util/setup_common.cpp:
```diff
@@ -184,8 +195,27 @@ BasicTestingSetup::BasicTestingSetup(const ChainType chainType, TestOpts opts)
+        // Traverse lower into /datadir
+        cached_dir = cached_dir / "datadir";
+
+        m_path_root = rand_path();
+        TryCreateDirectories(m_path_root);
+
+        // Copy the cached directory into the newly created temporary directory.
+        fs::copy(cached_dir, m_path_root, fs::copy_options::recursive);
```
When fuzzing on macOS (Apple Silicon), I ran into lots of crashes when copying the cached dir into the destination:

```
libc++abi: terminating due to uncaught exception of type std::__1::__fs::filesystem::filesystem_error: filesystem error: in copy_file: File exists ["/var/folders/wc/3l6l_0b92zzg54v6rmvqqn300000gn/T/test_common bitcoin/cmpctblock/6c5e38c6dcfbc2607321/regtest/blocks/xor.dat"] ["/var/folders/wc/3l6l_0b92zzg54v6rmvqqn300000gn/T/cmpctblock_cachedca12974312651650d6b0/datadir/regtest/blocks/xor.dat"]
```

The issue is that the destination dir may already exist before `TryCreateDirectories()`, and it happens when the AFL++ fork server is enabled (it does not happen when fuzzing with `AFL_NO_FORKSRV=1`). It appears that forked processes are running into collisions when generating the random path (`rand_path()`) for creating the destination dir, and when we switch to strong randomness (like how the static dir name is generated), the issue seems to be resolved:
```diff
diff --git a/src/test/util/setup_common.cpp b/src/test/util/setup_common.cpp
index e9768d4c5c..ac477caa62 100644
--- a/src/test/util/setup_common.cpp
+++ b/src/test/util/setup_common.cpp
@@ -164,8 +164,9 @@ BasicTestingSetup::BasicTestingSetup(const ChainType chainType, TestOpts opts)
         // tests, such as the fuzz tests to run in several processes at the
         // same time, add a random element to the path. Keep it small enough to
         // avoid a MAX_PATH violation on Windows.
-        const auto rand{HexStr(g_rng_temp_path.randbytes(10))};
-        return fs::temp_directory_path() / TEST_DIR_PATH_ELEMENT / test_name / rand;
+        std::vector<unsigned char> random_path_suffix(10);
+        GetStrongRandBytes(random_path_suffix);
+        return fs::temp_directory_path() / TEST_DIR_PATH_ELEMENT / test_name / HexStr(random_path_suffix);
     };
 
     if (m_node.args->IsArgSet("-testdatadir")) {
```

Crypt-iQ replied:
Are you fuzzing with persistent mode?

Crypt-iQ replied:
Think this is because the child processes have the same `g_rng_temp_path` so they create the same directory. Will need to see if `GetStrongRandBytes` introduces non-determinism. I'm confused why fuzzing on debian didn't bring this up.

marcofleon commented on src/test/fuzz/cmpctblock.cpp:
```diff
@@ -0,0 +1,422 @@
+
+    std::vector<CNode*> peers;
+    auto& connman = *static_cast<ConnmanTestMsg*>(setup->m_node.connman.get());
+    for (int i = 0; i < 3; ++i) {
```
As we discussed a bit offline, increasing the number of peers here to greater than 3 should be enough to hit some of the missing "eviction" logic in `MaybeSetPeerAsAnnouncingHeaderAndIDs`. Maybe increasing to 8 peers would be fine? Although not sure if that would test anything more (somewhere else) than if it were 4.

marcofleon commented on src/node/blockstorage.cpp:
```diff
@@ -487,7 +492,7 @@ bool BlockManager::WriteBlockIndexDB()
     std::vector<const CBlockIndex*> vBlocks;
     vBlocks.reserve(m_dirty_blockindex.size());
-    for (std::set<CBlockIndex*>::iterator it = m_dirty_blockindex.begin(); it != m_dirty_blockindex.end();) {
+    for (std::set<CBlockIndex*, CBlockIndexBlockHashComparator>::iterator it = m_dirty_blockindex.begin(); it != m_dirty_blockindex.end();) {
```
I feel like I might be overlooking something here, but could we keep `m_dirty_blockindex` the same and then if `EnableFuzzDeterminism()` just sort `vBlocks` by hash before calling `WriteBatchSync()`?

I guess I'm asking is there a reason why `m_dirty_blockindex` needs to be sorted by hash from the beginning vs right before the write.

Crypt-iQ replied:
> Is my understanding correct that we want to remove useless jitter that doesn't actually increase code coverage but confuses the fuzzer, such as the internal Random in leveldb's skiplist?

It depends on the specific fuzz engine as far as _exactly_ what happens (e.g. how it prioritizes or drops executed inputs). Speaking generally, if the fuzzer sees increased coverage (due to non-determinisim) it could prioritize a no-coverage-gain input and if it sees decreased coverage it could de-prioritize an input where it shouldn't.

To clarify if I was unclear above, this non-determinism doesn't show up in what hits the disk (as it's a k-v store), only in the `MemTable` which depends on the order that we write to it.

> I think we should tackle these in a separate PR.

This sounds fine to me and I can drop the commit here. https://github.com/bitcoin/bitcoin/pull/33469 is another case where a set of pointers caused non-determinism during fuzzing.

<details>
<summary> contrib/devtools/deterministic-fuzz-coverage output without this commit </summary>

```
diff --git a/Users/eugenesiegel/btc/bitcoin/build_fuzzcov/fuzz_det_cov.show.t1.a.txt b/Users/eugenesiegel/btc/bitcoin/build_fuzzcov/fuzz_det_cov.show.t1.b.txt
index 8389ec5527..68708cc9fc 100644
--- a/Users/eugenesiegel/btc/bitcoin/build_fuzzcov/fuzz_det_cov.show.t1.a.txt
+++ b/Users/eugenesiegel/btc/bitcoin/build_fuzzcov/fuzz_det_cov.show.t1.b.txt
@@ -61402,15 +61402,15 @@
    45|      0|  return "leveldb.InternalKeyComparator";
    46|      0|}
    47|       |
-   48|  6.11k|int InternalKeyComparator::Compare(const Slice& akey, const Slice& bkey) const {
+   48|  6.01k|int InternalKeyComparator::Compare(const Slice& akey, const Slice& bkey) const {
    49|       |  // Order by:
    50|       |  //    increasing user key (according to user-supplied comparator)
    51|       |  //    decreasing sequence number
    52|       |  //    decreasing type (though sequence# should be enough to disambiguate)
-   53|  6.11k|  int r = user_comparator_->Compare(ExtractUserKey(akey), ExtractUserKey(bkey));
-   54|  6.11k|  if (r == 0) {
+   53|  6.01k|  int r = user_comparator_->Compare(ExtractUserKey(akey), ExtractUserKey(bkey));
+   54|  6.01k|  if (r == 0) {
   ------------------
-  |  Branch (54:7): [True: 61, False: 6.05k]
+  |  Branch (54:7): [True: 61, False: 5.95k]
   ------------------
    55|     61|    const uint64_t anum = DecodeFixed64(akey.data() + akey.size() - 8);
    56|     61|    const uint64_t bnum = DecodeFixed64(bkey.data() + bkey.size() - 8);
@@ -61426,8 +61426,8 @@
    60|     51|      r = +1;
    61|     51|    }
    62|     61|  }
-   63|  6.11k|  return r;
-   64|  6.11k|}
+   63|  6.01k|  return r;
+   64|  6.01k|}
    65|       |
    66|       |void InternalKeyComparator::FindShortestSeparator(std::string* start,
    67|     10|                                                  const Slice& limit) const {
@@ -61620,10 +61620,10 @@
    92|       |bool ParseInternalKey(const Slice& internal_key, ParsedInternalKey* result);
    93|       |
    94|       |// Returns the user key portion of an internal key.
-   95|  13.1k|inline Slice ExtractUserKey(const Slice& internal_key) {
-   96|  13.1k|  assert(internal_key.size() >= 8);
-   97|  13.1k|  return Slice(internal_key.data(), internal_key.size() - 8);
-   98|  13.1k|}
+   95|  12.9k|inline Slice ExtractUserKey(const Slice& internal_key) {
+   96|  12.9k|  assert(internal_key.size() >= 8);
+   97|  12.9k|  return Slice(internal_key.data(), internal_key.size() - 8);
+   98|  12.9k|}
    99|       |
   100|       |// A comparator for internal keys that uses a specified comparator for
   101|       |// the user key portion and breaks ties by decreasing sequence number.
@@ -62498,12 +62498,12 @@
    11|       |
    12|       |namespace leveldb {
    13|       |
-   14|  11.1k|static Slice GetLengthPrefixedSlice(const char* data) {
-   15|  11.1k|  uint32_t len;
-   16|  11.1k|  const char* p = data;
-   17|  11.1k|  p = GetVarint32Ptr(p, p + 5, &len);  // +5: we assume "p" is not corrupted
-   18|  11.1k|  return Slice(p, len);
-   19|  11.1k|}
+   14|  10.9k|static Slice GetLengthPrefixedSlice(const char* data) {
+   15|  10.9k|  uint32_t len;
+   16|  10.9k|  const char* p = data;
+   17|  10.9k|  p = GetVarint32Ptr(p, p + 5, &len);  // +5: we assume "p" is not corrupted
+   18|  10.9k|  return Slice(p, len);
+   19|  10.9k|}
    20|       |
    21|       |MemTable::MemTable(const InternalKeyComparator& comparator)
    22|      4|    : comparator_(comparator), refs_(0), table_(comparator_, &arena_) {}
@@ -62513,12 +62513,12 @@
    26|      5|size_t MemTable::ApproximateMemoryUsage() { return arena_.MemoryUsage(); }
    27|       |
    28|       |int MemTable::KeyComparator::operator()(const char* aptr,
-   29|  4.93k|                                        const char* bptr) const {
+   29|  4.83k|                                        const char* bptr) const {
    30|       |  // Internal keys are encoded as length-prefixed strings.
-   31|  4.93k|  Slice a = GetLengthPrefixedSlice(aptr);
-   32|  4.93k|  Slice b = GetLengthPrefixedSlice(bptr);
-   33|  4.93k|  return comparator.Compare(a, b);
-   34|  4.93k|}
+   31|  4.83k|  Slice a = GetLengthPrefixedSlice(aptr);
+   32|  4.83k|  Slice b = GetLengthPrefixedSlice(bptr);
+   33|  4.83k|  return comparator.Compare(a, b);
+   34|  4.83k|}
    35|       |
    36|       |// Encode a suitable internal key target for "target" and return it.
    37|       |// Uses *scratch as scratch space, and the returned pointer will point
@@ -62883,12 +62883,12 @@
   150|       |
   151|       |  // Accessors/mutators for links.  Wrapped in methods so we can
   152|       |  // add the appropriate barriers as necessary.
-  153|  5.55k|  Node* Next(int n) {
-  154|  5.55k|    assert(n >= 0);
+  153|  5.24k|  Node* Next(int n) {
+  154|  5.24k|    assert(n >= 0);
   155|       |    // Use an 'acquire load' so that we observe a fully initialized
   156|       |    // version of the returned Node.
-  157|  5.55k|    return next_[n].load(std::memory_order_acquire);
-  158|  5.55k|  }
+  157|  5.24k|    return next_[n].load(std::memory_order_acquire);
+  158|  5.24k|  }
   159|    587|  void SetNext(int n, Node* x) {
   160|    587|    assert(n >= 0);
   161|       |    // Use a 'release store' so that anybody who reads through this
@@ -62995,15 +62995,15 @@
   252|    414|}
   253|       |
   254|       |template <typename Key, class Comparator>
-  255|  5.13k|bool SkipList<Key, Comparator>::KeyIsAfterNode(const Key& key, Node* n) const {
+  255|  4.82k|bool SkipList<Key, Comparator>::KeyIsAfterNode(const Key& key, Node* n) const {
   256|       |  // null n is considered infinite
-  257|  5.13k|  return (n != nullptr) && (compare_(n->key, key) < 0);
-                                         ^4.52k
+  257|  4.82k|  return (n != nullptr) && (compare_(n->key, key) < 0);
+                                         ^4.42k
   ------------------
-  |  Branch (257:10): [True: 4.52k, False: 613]
-  |  Branch (257:28): [True: 3.25k, False: 1.26k]
+  |  Branch (257:10): [True: 4.42k, False: 403]
+  |  Branch (257:28): [True: 2.94k, False: 1.47k]
   ------------------
-  258|  5.13k|}
+  258|  4.82k|}
   259|       |
   260|       |template <typename Key, class Comparator>
   261|       |typename SkipList<Key, Comparator>::Node*
@@ -63011,18 +63011,18 @@
   263|    443|                                              Node** prev) const {
   264|    443|  Node* x = head_;
   265|    443|  int level = GetMaxHeight() - 1;
-  266|  5.13k|  while (true) {
+  266|  4.82k|  while (true) {
   ------------------
   |  Branch (266:10): [Folded - Ignored]
   ------------------
-  267|  5.13k|    Node* next = x->Next(level);
-  268|  5.13k|    if (KeyIsAfterNode(key, next)) {
+  267|  4.82k|    Node* next = x->Next(level);
+  268|  4.82k|    if (KeyIsAfterNode(key, next)) {
   ------------------
-  |  Branch (268:9): [True: 3.25k, False: 1.87k]
+  |  Branch (268:9): [True: 2.94k, False: 1.87k]
   ------------------
   269|       |      // Keep searching in this list
-  270|  3.25k|      x = next;
-  271|  3.25k|    } else {
+  270|  2.94k|      x = next;
+  271|  2.94k|    } else {
   272|  1.87k|      if (prev != nullptr) prev[level] = x;
                                          ^1.85k
   ------------------
@@ -63038,7 +63038,7 @@
   277|  1.43k|        level--;
   278|  1.43k|      }
   279|  1.87k|    }
-  280|  5.13k|  }
+  280|  4.82k|  }
   281|    443|}
   282|       |
   283|       |template <typename Key, class Comparator>
@@ -68240,7 +68240,7 @@
    31|    945|  Slice() : data_(""), size_(0) {}
    32|       |
    33|       |  // Create a slice that refers to d[0,n-1].
-   34|  27.6k|  Slice(const char* d, size_t n) : data_(d), size_(n) {}
+   34|  27.2k|  Slice(const char* d, size_t n) : data_(d), size_(n) {}
    35|       |
    36|       |  // Create a slice that refers to the contents of "s"
    37|  2.16k|  Slice(const std::string& s) : data_(s.data()), size_(s.size()) {}
@@ -68253,10 +68253,10 @@
    44|       |  Slice& operator=(const Slice&) = default;
    45|       |
    46|       |  // Return a pointer to the beginning of the referenced data
-   47|  22.1k|  const char* data() const { return data_; }
+   47|  21.9k|  const char* data() const { return data_; }
    48|       |
    49|       |  // Return the length (in bytes) of the referenced data
-   50|  40.5k|  size_t size() const { return size_; }
+   50|  40.1k|  size_t size() const { return size_; }
    51|       |
    52|       |  // Return true iff the length of the referenced data is zero
    53|    433|  bool empty() const { return size_ == 0; }
@@ -68322,16 +68322,16 @@
   101|       |
   102|     26|inline bool operator!=(const Slice& x, const Slice& y) { return !(x == y); }
   103|       |
-  104|  6.41k|inline int Slice::compare(const Slice& b) const {
-  105|  6.41k|  const size_t min_len = (size_ < b.size_) ? size_ : b.size_;
-                                                           ^227    ^6.19k
+  104|  6.31k|inline int Slice::compare(const Slice& b) const {
+  105|  6.31k|  const size_t min_len = (size_ < b.size_) ? size_ : b.size_;
+                                                           ^227    ^6.09k
   ------------------
-  |  Branch (105:26): [True: 227, False: 6.19k]
+  |  Branch (105:26): [True: 227, False: 6.09k]
   ------------------
-  106|  6.41k|  int r = memcmp(data_, b.data_, min_len);
-  107|  6.41k|  if (r == 0) {
+  106|  6.31k|  int r = memcmp(data_, b.data_, min_len);
+  107|  6.31k|  if (r == 0) {
   ------------------
-  |  Branch (107:7): [True: 89, False: 6.32k]
+  |  Branch (107:7): [True: 89, False: 6.22k]
   ------------------
   108|     89|    if (size_ < b.size_)
   ------------------
@@ -68344,8 +68344,8 @@
   ------------------
   111|      0|      r = +1;
   112|     89|  }
-  113|  6.41k|  return r;
-  114|  6.41k|}
+  113|  6.31k|  return r;
+  114|  6.31k|}
   115|       |
   116|       |}  // namespace leveldb
   117|       |
@@ -72351,22 +72351,22 @@
   106|       |const char* GetVarint32PtrFallback(const char* p, const char* limit,
   107|       |                                   uint32_t* value);
   108|       |inline const char* GetVarint32Ptr(const char* p, const char* limit,
-  109|  11.9k|                                  uint32_t* value) {
-  110|  11.9k|  if (p < limit) {
+  109|  11.7k|                                  uint32_t* value) {
+  110|  11.7k|  if (p < limit) {
   ------------------
-  |  Branch (110:7): [True: 11.9k, False: 4]
+  |  Branch (110:7): [True: 11.7k, False: 4]
   ------------------
-  111|  11.9k|    uint32_t result = *(reinterpret_cast<const uint8_t*>(p));
-  112|  11.9k|    if ((result & 128) == 0) {
+  111|  11.7k|    uint32_t result = *(reinterpret_cast<const uint8_t*>(p));
diff --git a/Users/eugenesiegel/btc/bitcoin/build_fuzzcov/fuzz_det_cov.show.t0.a.txt b/Users/eugenesiegel/btc/bitcoin/build_fuzzcov/fuzz_det_cov.show.t0.b.txt
index 209490c47e..67c0258d57 100644
+  112|  11.7k|    if ((result & 128) == 0) {
   ------------------
-  |  Branch (112:9): [True: 11.9k, False: 0]
--- a/Users/eugenesiegel/btc/bitcoin/build_fuzzcov/fuzz_det_cov.show.t0.a.txt
+  |  Branch (112:9): [True: 11.7k, False: 0]
+++ b/Users/eugenesiegel/btc/bitcoin/build_fuzzcov/fuzz_det_cov.show.t0.b.txt
   ------------------
-  113|  11.9k|      *value = result;
-  114|  11.9k|      return p + 1;
-  115|  11.9k|    }
-  116|  11.9k|  }
@@ -61393,15 +61393,15 @@
+  113|  11.7k|      *value = result;
+  114|  11.7k|      return p + 1;
    45|      0|  return "leveldb.InternalKeyComparator";
+  115|  11.7k|    }
    46|      0|}
+  116|  11.7k|  }
    47|       |
   117|      4|  return GetVarint32PtrFallback(p, limit, value);
-  118|  11.9k|}
+  118|  11.7k|}
-   48|  6.06k|int InternalKeyComparator::Compare(const Slice& akey, const Slice& bkey) const {
   119|       |
   120|       |}  // namespace leveldb
   121|       |
+   48|  6.01k|int InternalKeyComparator::Compare(const Slice& akey, const Slice& bkey) const {
@@ -72399,9 +72399,9 @@
    24|       |
    25|      4|  const char* Name() const override { return "leveldb.BytewiseComparator"; }
    26|       |
    49|       |  // Order by:
-   27|  6.41k|  int Compare(const Slice& a, const Slice& b) const override {
    50|       |  //    increasing user key (according to user-supplied comparator)
-   28|  6.41k|    return a.compare(b);
    51|       |  //    decreasing sequence number
-   29|  6.41k|  }
    52|       |  //    decreasing type (though sequence# should be enough to disambiguate)
+   27|  6.31k|  int Compare(const Slice& a, const Slice& b) const override {
-   53|  6.06k|  int r = user_comparator_->Compare(ExtractUserKey(akey), ExtractUserKey(bkey));
+   28|  6.31k|    return a.compare(b);
-   54|  6.06k|  if (r == 0) {
+   29|  6.31k|  }
+   53|  6.01k|  int r = user_comparator_->Compare(ExtractUserKey(akey), ExtractUserKey(bkey));
    30|       |
+   54|  6.01k|  if (r == 0) {
    31|       |  void FindShortestSeparator(std::string* start,
   ------------------
    32|     10|                             const Slice& limit) const override {
-  |  Branch (54:7): [True: 66, False: 5.99k]
+  |  Branch (54:7): [True: 66, False: 5.94k]
   ------------------
    55|     66|    const uint64_t anum = DecodeFixed64(akey.data() + akey.size() - 8);
    56|     66|    const uint64_t bnum = DecodeFixed64(bkey.data() + bkey.size() - 8);
@@ -61417,8 +61417,8 @@
    60|     56|      r = +1;
    61|     56|    }
    62|     66|  }
-   63|  6.06k|  return r;
-   64|  6.06k|}
+   63|  6.01k|  return r;
+   64|  6.01k|}
    65|       |
    66|       |void InternalKeyComparator::FindShortestSeparator(std::string* start,
    67|     10|                                                  const Slice& limit) const {
@@ -61611,10 +61611,10 @@
    92|       |bool ParseInternalKey(const Slice& internal_key, ParsedInternalKey* result);
    93|       |
    94|       |// Returns the user key portion of an internal key.
-   95|  13.0k|inline Slice ExtractUserKey(const Slice& internal_key) {
-   96|  13.0k|  assert(internal_key.size() >= 8);
-   97|  13.0k|  return Slice(internal_key.data(), internal_key.size() - 8);
-   98|  13.0k|}
+   95|  12.9k|inline Slice ExtractUserKey(const Slice& internal_key) {
+   96|  12.9k|  assert(internal_key.size() >= 8);
+   97|  12.9k|  return Slice(internal_key.data(), internal_key.size() - 8);
+   98|  12.9k|}
    99|       |
   100|       |// A comparator for internal keys that uses a specified comparator for
   101|       |// the user key portion and breaks ties by decreasing sequence number.
@@ -62489,12 +62489,12 @@
    11|       |
    12|       |namespace leveldb {
    13|       |
-   14|  10.9k|static Slice GetLengthPrefixedSlice(const char* data) {
-   15|  10.9k|  uint32_t len;
-   16|  10.9k|  const char* p = data;
-   17|  10.9k|  p = GetVarint32Ptr(p, p + 5, &len);  // +5: we assume "p" is not corrupted
-   18|  10.9k|  return Slice(p, len);
-   19|  10.9k|}
+   14|  10.8k|static Slice GetLengthPrefixedSlice(const char* data) {
+   15|  10.8k|  uint32_t len;
+   16|  10.8k|  const char* p = data;
+   17|  10.8k|  p = GetVarint32Ptr(p, p + 5, &len);  // +5: we assume "p" is not corrupted
+   18|  10.8k|  return Slice(p, len);
+   19|  10.8k|}
    20|       |
    21|       |MemTable::MemTable(const InternalKeyComparator& comparator)
    22|      4|    : comparator_(comparator), refs_(0), table_(comparator_, &arena_) {}
@@ -62504,12 +62504,12 @@
    26|      5|size_t MemTable::ApproximateMemoryUsage() { return arena_.MemoryUsage(); }
    27|       |
    28|       |int MemTable::KeyComparator::operator()(const char* aptr,
-   29|  4.83k|                                        const char* bptr) const {
+   29|  4.78k|                                        const char* bptr) const {
    30|       |  // Internal keys are encoded as length-prefixed strings.
-   31|  4.83k|  Slice a = GetLengthPrefixedSlice(aptr);
-   32|  4.83k|  Slice b = GetLengthPrefixedSlice(bptr);
-   33|  4.83k|  return comparator.Compare(a, b);
-   34|  4.83k|}
+   31|  4.78k|  Slice a = GetLengthPrefixedSlice(aptr);
+   32|  4.78k|  Slice b = GetLengthPrefixedSlice(bptr);
+   33|  4.78k|  return comparator.Compare(a, b);
+   34|  4.78k|}
    35|       |
    36|       |// Encode a suitable internal key target for "target" and return it.
    37|       |// Uses *scratch as scratch space, and the returned pointer will point
@@ -62874,12 +62874,12 @@
   150|       |
   151|       |  // Accessors/mutators for links.  Wrapped in methods so we can
   152|       |  // add the appropriate barriers as necessary.
-  153|  5.37k|  Node* Next(int n) {
-  154|  5.37k|    assert(n >= 0);
+  153|  5.14k|  Node* Next(int n) {
+  154|  5.14k|    assert(n >= 0);
   155|       |    // Use an 'acquire load' so that we observe a fully initialized
   156|       |    // version of the returned Node.
-  157|  5.37k|    return next_[n].load(std::memory_order_acquire);
-  158|  5.37k|  }
+  157|  5.14k|    return next_[n].load(std::memory_order_acquire);
+  158|  5.14k|  }
   159|    587|  void SetNext(int n, Node* x) {
   160|    587|    assert(n >= 0);
   161|       |    // Use a 'release store' so that anybody who reads through this
@@ -62986,15 +62986,15 @@
   252|    414|}
   253|       |
   254|       |template <typename Key, class Comparator>
-  255|  4.95k|bool SkipList<Key, Comparator>::KeyIsAfterNode(const Key& key, Node* n) const {
+  255|  4.73k|bool SkipList<Key, Comparator>::KeyIsAfterNode(const Key& key, Node* n) const {
   256|       |  // null n is considered infinite
-  257|  4.95k|  return (n != nullptr) && (compare_(n->key, key) < 0);
-                                         ^4.42k
+  257|  4.73k|  return (n != nullptr) && (compare_(n->key, key) < 0);
+                                         ^4.37k
   ------------------
-  |  Branch (257:10): [True: 4.42k, False: 534]
-  |  Branch (257:28): [True: 3.07k, False: 1.35k]
+  |  Branch (257:10): [True: 4.37k, False: 356]
+  |  Branch (257:28): [True: 2.84k, False: 1.52k]
   ------------------
-  258|  4.95k|}
+  258|  4.73k|}
   259|       |
   260|       |template <typename Key, class Comparator>
   261|       |typename SkipList<Key, Comparator>::Node*
@@ -63002,18 +63002,18 @@
   263|    448|                                              Node** prev) const {
   264|    448|  Node* x = head_;
   265|    448|  int level = GetMaxHeight() - 1;
-  266|  4.95k|  while (true) {
+  266|  4.73k|  while (true) {
   ------------------
   |  Branch (266:10): [Folded - Ignored]
   ------------------
-  267|  4.95k|    Node* next = x->Next(level);
-  268|  4.95k|    if (KeyIsAfterNode(key, next)) {
+  267|  4.73k|    Node* next = x->Next(level);
+  268|  4.73k|    if (KeyIsAfterNode(key, next)) {
   ------------------
-  |  Branch (268:9): [True: 3.07k, False: 1.88k]
+  |  Branch (268:9): [True: 2.84k, False: 1.88k]
   ------------------
   269|       |      // Keep searching in this list
-  270|  3.07k|      x = next;
-  271|  3.07k|    } else {
+  270|  2.84k|      x = next;
+  271|  2.84k|    } else {
   272|  1.88k|      if (prev != nullptr) prev[level] = x;
                                          ^1.85k
   ------------------
@@ -63029,7 +63029,7 @@
   277|  1.43k|        level--;
   278|  1.43k|      }
   279|  1.88k|    }
-  280|  4.95k|  }
+  280|  4.73k|  }
   281|    448|}
   282|       |
   283|       |template <typename Key, class Comparator>
@@ -68231,7 +68231,7 @@
    31|    985|  Slice() : data_(""), size_(0) {}
    32|       |
    33|       |  // Create a slice that refers to d[0,n-1].
-   34|  27.4k|  Slice(const char* d, size_t n) : data_(d), size_(n) {}
+   34|  27.2k|  Slice(const char* d, size_t n) : data_(d), size_(n) {}
    35|       |
    36|       |  // Create a slice that refers to the contents of "s"
    37|  2.20k|  Slice(const std::string& s) : data_(s.data()), size_(s.size()) {}
@@ -68244,10 +68244,10 @@
    44|       |  Slice& operator=(const Slice&) = default;
    45|       |
    46|       |  // Return a pointer to the beginning of the referenced data
-   47|  22.1k|  const char* data() const { return data_; }
+   47|  22.0k|  const char* data() const { return data_; }
    48|       |
    49|       |  // Return the length (in bytes) of the referenced data
-   50|  40.4k|  size_t size() const { return size_; }
+   50|  40.2k|  size_t size() const { return size_; }
    51|       |
    52|       |  // Return true iff the length of the referenced data is zero
    53|    433|  bool empty() const { return size_ == 0; }
@@ -68313,16 +68313,16 @@
   101|       |
   102|     31|inline bool operator!=(const Slice& x, const Slice& y) { return !(x == y); }
   103|       |
-  104|  6.37k|inline int Slice::compare(const Slice& b) const {
-  105|  6.37k|  const size_t min_len = (size_ < b.size_) ? size_ : b.size_;
-                                                           ^250    ^6.12k
+  104|  6.32k|inline int Slice::compare(const Slice& b) const {
+  105|  6.32k|  const size_t min_len = (size_ < b.size_) ? size_ : b.size_;
+                                                           ^239    ^6.08k
   ------------------
-  |  Branch (105:26): [True: 250, False: 6.12k]
+  |  Branch (105:26): [True: 239, False: 6.08k]
   ------------------
-  106|  6.37k|  int r = memcmp(data_, b.data_, min_len);
-  107|  6.37k|  if (r == 0) {
+  106|  6.32k|  int r = memcmp(data_, b.data_, min_len);
+  107|  6.32k|  if (r == 0) {
   ------------------
-  |  Branch (107:7): [True: 99, False: 6.27k]
+  |  Branch (107:7): [True: 99, False: 6.22k]
   ------------------
   108|     99|    if (size_ < b.size_)
   ------------------
@@ -68335,8 +68335,8 @@
   ------------------
   111|      0|      r = +1;
   112|     99|  }
-  113|  6.37k|  return r;
-  114|  6.37k|}
+  113|  6.32k|  return r;
+  114|  6.32k|}
   115|       |
   116|       |}  // namespace leveldb
   117|       |
@@ -71306,15 +71306,15 @@
    45|    418|  char* result;
    46|    418|  if (needed <= alloc_bytes_remaining_) {
   ------------------
-  |  Branch (46:7): [True: 411, False: 7]
+  |  Branch (46:7): [True: 412, False: 6]
   ------------------
-   47|    411|    result = alloc_ptr_ + slop;
-   48|    411|    alloc_ptr_ += needed;
-   49|    411|    alloc_bytes_remaining_ -= needed;
-   50|    411|  } else {
+   47|    412|    result = alloc_ptr_ + slop;
+   48|    412|    alloc_ptr_ += needed;
+   49|    412|    alloc_bytes_remaining_ -= needed;
+   50|    412|  } else {
    51|       |    // AllocateFallback always returned aligned memory
-   52|      7|    result = AllocateFallback(bytes);
-   53|      7|  }
+   52|      6|    result = AllocateFallback(bytes);
+   53|      6|  }
    54|    418|  assert((reinterpret_cast<uintptr_t>(result) & (align - 1)) == 0);
    55|    418|  return result;
    56|    418|}
@@ -71391,14 +71391,14 @@
    59|    414|  assert(bytes > 0);
    60|    414|  if (bytes <= alloc_bytes_remaining_) {
   ------------------
-  |  Branch (60:7): [True: 405, False: 9]
+  |  Branch (60:7): [True: 404, False: 10]
   ------------------
-   61|    405|    char* result = alloc_ptr_;
-   62|    405|    alloc_ptr_ += bytes;
-   63|    405|    alloc_bytes_remaining_ -= bytes;
-   64|    405|    return result;
-   65|    405|  }
-   66|      9|  return AllocateFallback(bytes);
+   61|    404|    char* result = alloc_ptr_;
+   62|    404|    alloc_ptr_ += bytes;
+   63|    404|    alloc_bytes_remaining_ -= bytes;
+   64|    404|    return result;
+   65|    404|  }
+   66|     10|  return AllocateFallback(bytes);
    67|    414|}
    68|       |
    69|       |}  // namespace leveldb
@@ -72342,22 +72342,22 @@
   106|       |const char* GetVarint32PtrFallback(const char* p, const char* limit,
   107|       |                                   uint32_t* value);
   108|       |inline const char* GetVarint32Ptr(const char* p, const char* limit,
-  109|  11.7k|                                  uint32_t* value) {
-  110|  11.7k|  if (p < limit) {
+  109|  11.6k|                                  uint32_t* value) {
+  110|  11.6k|  if (p < limit) {
   ------------------
-  |  Branch (110:7): [True: 11.7k, False: 4]
+  |  Branch (110:7): [True: 11.6k, False: 4]
   ------------------
-  111|  11.7k|    uint32_t result = *(reinterpret_cast<const uint8_t*>(p));
-  112|  11.7k|    if ((result & 128) == 0) {
+  111|  11.6k|    uint32_t result = *(reinterpret_cast<const uint8_t*>(p));
+  112|  11.6k|    if ((result & 128) == 0) {
   ------------------
-  |  Branch (112:9): [True: 11.7k, False: 0]
+  |  Branch (112:9): [True: 11.6k, False: 0]
   ------------------
-  113|  11.7k|      *value = result;
-  114|  11.7k|      return p + 1;
-  115|  11.7k|    }
-  116|  11.7k|  }
+  113|  11.6k|      *value = result;
+  114|  11.6k|      return p + 1;
+  115|  11.6k|    }
+  116|  11.6k|  }
   117|      4|  return GetVarint32PtrFallback(p, limit, value);
-  118|  11.7k|}
+  118|  11.6k|}
   119|       |
   120|       |}  // namespace leveldb
   121|       |
@@ -72390,9 +72390,9 @@
    24|       |
    25|      4|  const char* Name() const override { return "leveldb.BytewiseComparator"; }
    26|       |
-   27|  6.36k|  int Compare(const Slice& a, const Slice& b) const override {
-   28|  6.36k|    return a.compare(b);
-   29|  6.36k|  }
+   27|  6.32k|  int Compare(const Slice& a, const Slice& b) const override {
+   28|  6.32k|    return a.compare(b);
+   29|  6.32k|  }
    30|       |
    31|       |  void FindShortestSeparator(std::string* start,
    32|     10|                             const Slice& limit) const override {
⚠️

The coverage was not deterministic between runs.
The fuzz target input was /Users/eugenesiegel/btc/bitcoin/cmpctblock/cmpctblock/00095f04def2d4e203a2844032ee4bb88e33681d
```
</details>

marcofleon reviewed (commented):
Some possible discussion about ed813c48f826d083becf93c741b483774c850c86 for review club.

l0rinc commented on src/test/fuzz/util.h:
```diff
@@ -340,4 +342,11 @@ void ReadFromStream(FuzzedDataProvider& fuzzed_data_provider, Stream& stream) no
     }
 }
 
+inline void FinalizeHeader(CBlockHeader& header, const ChainstateManager& chainman)
```
nit: please mention in the commit message why the method was slightly changed besides the move

l0rinc reviewed (changes requested):
Concept and approach ACK, thanks for tackling this.

I like how you've extracted the first few commits that I could quickly get out of the way to continue to the more difficult ones - which are a bit too difficult though, it would help me personally if we could split it into smaller functional chunks, so that the reviewers are guided through the change. The commit messages could give extra context where needed. I don't think we should split by functions, but rather functionalities so that each commit could theoretically be merged if needed - we shouldn't depend on a future commit to understand a previous one.

I have provided some hints and some further context for the `LevelDB` related changes. I think we should do the set stabilization in a dedicated PR, but you can also adjust it here so that it likely performs similarly to the original.

Crypt-iQ commented on src/test/fuzz/cmpctblock.cpp:
```diff
@@ -0,0 +1,429 @@
+                CBlockHeaderAndShortTxIDs base_cmpctblock = cmpctblock;
+                net_msg = NetMsg::Make(NetMsgType::CMPCTBLOCK, base_cmpctblock);
+            },
+            [&]() {
```
I'll incrementally add the cases so it's a bit easier to digest.

l0rinc commented on src/node/blockstorage.cpp:
```diff
@@ -180,6 +180,11 @@ bool CBlockIndexHeightOnlyComparator::operator()(const CBlockIndex* pa, const CB
 
+bool CBlockIndexBlockHashComparator::operator()(const CBlockIndex* pa, const CBlockIndex* pb) const
+{
+    return UintToArith256(pa->GetBlockHash()) < UintToArith256(pb->GetBlockHash());
```
The suggested comparator is a pure function, I don't see how it could be non-deterministic - unless it's broken, in which case we should definitely dig deeper.
Can you try it again an report what you see?

Crypt-iQ commented on src/test/util/setup_common.cpp:
```diff
@@ -287,6 +287,8 @@ ChainTestingSetup::ChainTestingSetup(const ChainType chainType, TestOpts opts)
             promise.get_future().wait();
         }
+    } else if (opts.setup_validation_interface_no_scheduler) {
+        m_node.validation_signals = std::make_unique<ValidationSignals>(std::make_unique<util::ImmediateTaskRunner>());
```
Done in 0190ac6622ff017aeead385289918d388a85218a

Crypt-iQ commented on src/test/fuzz/cmpctblock.cpp:
```diff
@@ -0,0 +1,429 @@
+        in.scriptWitness.stack = script_wit_stack;
+        tx_mut.vin.push_back(in);
+
+        const CAmount amount_in = GetAmount(outpoint);
```
Done in 04d1dc78599362623bdd4b55c7be0c7b0f9f2174

frankomosh reviewed (commented):
Concept ACK. fuzzing direction look reasonable.

Crypt-iQ commented on src/util/fs.h:
```diff
@@ -136,6 +136,13 @@ static inline bool copy_file(const path& from, const path& to, copy_options opti
 
+// Disallow implicit std::string conversion for copy to avoid locale-dependent
+// encoding on Windows. This is currently only used in fuzzing.
+static inline void copy(const path& from, const path& to, copy_options options)
```
This needs to be here because of the linter which checks that `std::filesystem` isn't used outside of this `fs` namespace.

Crypt-iQ committed:
fuzz: initial compact block fuzz harness

Adds a fuzz harness cmpctblock to test BIP152 compact block relay.
Currently just sets mock time in a loop.

Crypt-iQ committed:
net, fuzz: move CMPCTBLOCK_VERSION to header, use in cmpctblock harness

The cmpctblock harness now adds peers and processes the sendcmpct message.

Crypt-iQ committed:
fuzz: create and send transactions in cmpctblock harness

If the mempool is modified at all (determined by a change in the sequence
counter), reset the rng, mempool, and the chainman for the next iteration.

Crypt-iQ committed:
fuzz: mine blocks and send headers for them in cmpctblock harness

The blocks may include mempool and non-mempool transactions. If the
block index was added to, reset the rng, mempool, and chainman. Also
move FinalizeHeader from p2p_headers_presync.cpp to util.h so that
the mining function can use it to create valid headers.

Crypt-iQ committed:
fuzz: send compact blocks in cmpctblock harness

Using an existing block or a newly created block, send a compact
block message to the target node. Some of the time, the harness
will prefill multiple transactions from the block. Adds a
FuzzedCBlockHeaderAndShortTxIDs class that's used to populate the
protected prefilledtxn and shorttxids fields.

Crypt-iQ committed:
fuzz: send blocktxn messages in cmpctblock harness

Sometimes, blindly take an existing block and choose random
transactions to request in a blocktxn message.

Crypt-iQ commented:
Latest push to 3c515102f4e83b4117f52d7d0fd41aac59740b16 gets rid of all the `fs` and `TestingSetup` changes. Now if the mempool changed at all (if sequence numbers changed) or the block index changed, it resets the rng (calls `MakeRandDeterministicDANGEROUS`, its `extern` which I think is acceptable), resets the mempool, the chainman, and mines blocks again to make it deterministic. This is faster than the old version (which is surprising to me) and the changes are nicely contained to the fuzz code. Will post coverage later (though nothing should have changed).

instagibbs commented on src/test/fuzz/cmpctblock.cpp:
```diff
@@ -75,14 +87,46 @@ FUZZ_TARGET(cmpctblock, .init = initialize_cmpctblock)
+            [&]() {
+                // Send a sendcmpct message, optionally setting hb mode.
+                bool hb = fuzzed_data_provider.ConsumeBool();
+                net_msg = NetMsg::Make(NetMsgType::SENDCMPCT, /*high_bandwidth=*/hb, /*version=*/CMPCTBLOCKS_VERSION);
```
09f547384aff90359b8a9e9a6f469b14683ffbab

not a ton of value, but we should also not crash/etc if the wrong version is sent?

instagibbs commented on src/test/fuzz/cmpctblock.cpp:
```diff
@@ -340,6 +340,32 @@ FUZZ_TARGET(cmpctblock, .init = initialize_cmpctblock)
+                std::shared_ptr<CBlock> cblock = block_info.block;
+
+                size_t num_txs = cblock->vtx.size();
+                if (num_txs > 1) {
```
3c515102f4e83b4117f52d7d0fd41aac59740b16

Sending a coinbase is fair play for the target, no? Why gate?

instagibbs reviewed (commented):
strong concept ACK on the target, I do wonder what issues we would like to shoot for. 

For instance, do we want to cover things like protocol-invalid messages, malformed compact blocks, invalid PoW, etc.

Seems clearly better than what we have now at least.

Crypt-iQ commented on src/test/fuzz/cmpctblock.cpp:
```diff
@@ -115,6 +135,54 @@ FUZZ_TARGET(cmpctblock, .init = initialize_cmpctblock)
+        // If the mempool is non-empty, choose a mempool outpoint. Otherwise, choose a coinbase.
+        COutPoint outpoint;
+        unsigned long mempool_size = mempool.size();
+        if (fuzzed_data_provider.ConsumeBool() && mempool_size != 0) {
```
Yes good point, I can change this. I wanted to limit invalidity here, but there's no good reason that we can't double-spend the coinbase outpoints.

Crypt-iQ replied:
Changed in c93df0e9d93dde68554721463a96c752457fef58. The following commit a165c3f3ff63faf3a2420444f1dd6a3032154d89 also adds a case to spend an outpoint from a created block.

instagibbs commented on src/test/fuzz/cmpctblock.cpp:
```diff
@@ -113,6 +136,50 @@ FUZZ_TARGET(cmpctblock, .init = initialize_cmpctblock)
+        tx_mut.version = fuzzed_data_provider.ConsumeBool() ? CTransaction::CURRENT_VERSION : TRUC_VERSION;
+        tx_mut.nLockTime = fuzzed_data_provider.ConsumeBool() ? 0 : fuzzed_data_provider.ConsumeIntegral<uint32_t>();
+
+        // If the mempool is non-empty, choose a mempool outpoint. Otherwise, choose a coinbase.
```
c93df0e9d93dde68554721463a96c752457fef58

nit: comment makes it sound unconditional

Crypt-iQ replied:
Changed to "If the mempool is non-empty, sometimes choose a mempool outpoint. Otherwise, choose a coinbase." in 651622432d2d43d6503412094ad8c7bf16487b62.

marcofleon commented on src/test/fuzz/cmpctblock.cpp:
```diff
@@ -0,0 +1,502 @@
+        }
+        assert(num_hb <= 3);
+
+        if (sent_sendcmpct && random_node.fSuccessfullyConnected && !random_node.fDisconnect) {
```
In `net_processing` I'm not seeing an `fSuccessfullyConnected` check for the `SENDCMPCT` message. There is a check [further down](https://github.com/bitcoin/bitcoin/blob/f17cd18d02b4bd6ff0aed03f04968ea52271ee70/src/net_processing.cpp#L4012), after `SENDCMPCT`. Do we need to gate on it in this harness? Dropping it would check the assertion for every peer regardless of a completed handshake, which I think is the correct behavior unless I'm missing something.

Crypt-iQ replied:
I don't think it's needed, will run without and see if it crashes.

Crypt-iQ replied:
Changed in 8c9a3fd0e87f5855e398a46b482ade4644dc3e92. It didn't crash after a few hours -- I think that's enough time to test as simply removing the (needed) `fDisconnect` check crashes within a minute.

instagibbs commented on src/test/fuzz/cmpctblock.cpp:
```diff
@@ -155,6 +170,14 @@ FUZZ_TARGET(cmpctblock, .init = initialize_cmpctblock)
             outpoint = COutPoint(tx->GetHash(), 0);
             amount_in = tx->vout[0].nValue;
+        } else if (fuzzed_data_provider.ConsumeBool() && info.size() != 0) {
+            // These blocks (and txs) may be invalid or not in the main chain.
```
a165c3f3ff63faf3a2420444f1dd6a3032154d89

nit: or the output already spent

Crypt-iQ replied:
Changed to "These blocks (and txs) may be invalid, use a spent output, or not be in the main chain." in 3c58efe2acf0e7fbf1daaadd8bff2e0cfd3761e1.

instagibbs commented on src/test/fuzz/cmpctblock.cpp:
```diff
@@ -180,6 +203,78 @@ FUZZ_TARGET(cmpctblock, .init = initialize_cmpctblock)
+        uint256 prev;
+        uint32_t height;
+
+        if (fuzzed_data_provider.ConsumeBool() || info.size() == 0) {
```
a165c3f3ff63faf3a2420444f1dd6a3032154d89

swap the checks to reduce unnecessary reading of fuzz input?

Crypt-iQ replied:
Changed in 3c58efe2acf0e7fbf1daaadd8bff2e0cfd3761e1. Also changed order of checks in `create_tx` in both 3c58efe2acf0e7fbf1daaadd8bff2e0cfd3761e1 and 651622432d2d43d6503412094ad8c7bf16487b62.

instagibbs reviewed (commented):
Confirmed that blockchain is moving forward even with non-trivial compact blocks (not just cb)

~~Can't seem to generate a mempool tx in my short run that actually enters the mempool. Might want to check coverage in depth to validate the state you think we should be hitting.~~

ok found mempool entries, just took a few more minutes to hit

marcofleon reviewed (commented):
Ran with AFL++ for a couple days and stability ended up at 98.53% (nice) and exec/s was ~100, which is reasonable for a harness that does as much as this one does. Maybe lowering the `LIMITED_WHILE` max would help with that, if we care.

Left a question.

frankomosh commented on src/test/fuzz/cmpctblock.cpp:
```diff
@@ -0,0 +1,502 @@
+
+void ResetChainmanAndMempool(TestingSetup& setup)
+{
+    SetMockTime(Params().GenesisBlock().Time());
```
nit: could we use `NodeClockContext` instead?..not sure if that pattern works well here, but I just noticed that most recent tests are moving away from `SetMockTime`

Crypt-iQ replied:
I will copy the pattern used in `process_message(s)`, which keeps `SetMockTime` usage in the reset function and uses the context at the top of the loop. It should be fine to also use another context here, but I'd rather keep the style the same.

Crypt-iQ replied:
Changed in 6cd480f62f69f170e4eb24ea0ccca97e996e5333.

frankomosh commented on src/test/fuzz/cmpctblock.cpp:
```diff
@@ -279,6 +296,50 @@ FUZZ_TARGET(cmpctblock, .init = initialize_cmpctblock)
+                    cmpctblock.EraseShortTxIDs(i - 1 - num_erased);
+                    ++num_erased;
+                }
+
```
The count assert is a good cheap check. Just wondering if it could be complemented with another one. I think the totals would still add up correctly even if `AddPrefilledTx` silently does nothing or `EraseShortTxIDs` is never called, so those cases would pass undetected. Could maybe also track `selected_for_prefill` indices during the loop and assert they appear in `prefilledtxn` afterwards?

Crypt-iQ replied:
How would 1) `AddPrefilledTx` do nothing, and 2) how would `EraseShortTxIDs` never get called? Is there a bug in the current code that you can see? I don't think the existing assert is even necessary, I've looked at the coverage.

dergoegge replied:
@frankomosh The way your comment is worded and Eugene's response make me think your comment was LLM generated. I appreciate that you try to contribute constructively, but just posting LLM output is not useful during review and wastes time.

frankomosh replied:
I had ran mutation-core against the prefill logic and found that mutants removing the `AddPrefilledTx` call or skipping `EraseShortTxIDs` both survived the existing assert, meaning the count check doesn't distinguish correct encoding from those broken states. That's the scenario I had in mind, not a bug in the current code. If the assert isn't needed at all as you say, that's also useful to know. Was simply giving a suggestion and not objecting to anything.

frankomosh replied:
Maybe should have mentioned that first hand

Crypt-iQ replied:
Could you please provide the mutant "diff" @frankomosh? When I remove the `AddPrefilledTx` call, I get a crash on the assertion and I want to know if I'm overlooking something.

frankomosh replied:
Sure. Here are the mutant diffs:

Removes the `AddPrefilledTx` call:
```diff
-                    cmpctblock.AddPrefilledTx(std::move(prefilledtx));
+                    
```

Skips `EraseShortTxIDs`:
```diff
-                    cmpctblock.EraseShortTxIDs(i - num_erased);
+                    
```

frankomosh replied:
Wait... I might have made a mistake because I was using `build_fuzz_nosan`, (which I think compiles out asserts?)

Crypt-iQ replied:
With `--preset=libfuzzer-nosan` I hit the assertion with both mutants

Crypt-iQ commented:
I pushed twice because I forgot to address one comment. Can see the diff [here](https://github.com/bitcoin/bitcoin/compare/14e153aab5731e151d0a408432366db84d922b4a..c8d688f41ccb218f301adde4120f2094d13b739a). Coverage from last month is [here](https://crypt-iq.github.io/fuzz_coverage_reports/03242026-cmpctblock-coverage-report/) -- the corpus was invalidated so I will run again.

Edit:
> Maybe lowering the LIMITED_WHILE max would help with that

This might help for large inputs, I can compare 1000 to 100 and see if it makes a difference.

instagibbs commented:
ACK c8d688f41ccb218f301adde4120f2094d13b739a

Reasonable to have this land and can be improved more in the future

> This might help for large inputs, I can compare 1000 to 100 and see if it makes a difference.

Did you check this?

Crypt-iQ commented:
> Did you check this?

Currently running, will report back.

Crypt-iQ commented:
I ran for 3 hours with 1000 vs 100 in the loop condition and the numbers were basically the same at around 100 execs/s. Not a perfect benchmark since I didn't run many trials, but I don't think the bottleneck is in the loop condition. I did profile a while back and I did notice a speedup when I disabled `CheckBlockIndex`.

marcofleon reviewed (commented):
ACK c8d688f41ccb218f301adde4120f2094d13b739a

Lgtm, I don't think the 100 exec/s should be a blocker and can be improved later if need be.
